ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PeerSimulator/Handlers.hs

```diff
@@ -35,19 +36,43 @@ import qualified Test.Consensus.BlockTree as BT
 import           Test.Consensus.BlockTree (BlockTree)
 import           Test.Consensus.Network.AnchoredFragment.Extras (intersectWith)
 import           Test.Consensus.PeerSimulator.ScheduledBlockFetchServer
-                     (BlockFetch (..))
+                     (BlockFetch (..), SendBlocks (..))
 import           Test.Consensus.PeerSimulator.ScheduledChainSyncServer
                      (FindIntersect (..),
                      RequestNext (AwaitReply, RollBackward, RollForward))
-import           Test.Consensus.PeerSimulator.Trace (terseFrag, tersePoint)
+import           Test.Consensus.PeerSimulator.Trace (terseFrag, tersePoint, terseBlock)
 import           Test.Consensus.PointSchedule
 import           Test.Util.Orphans.IOLike ()
 import           Test.Util.TestBlock (TestBlock, TestHash)
+import Test.Util.TestBlock (TestHash(TestHash))
+import Ouroboros.Network.AnchoredFragment (AnchoredFragment)
+import qualified Data.List.NonEmpty as NonEmpty
 
 
 toPoint :: (HasHeader block, HeaderHash block ~ TestHash) => WithOrigin block -> Point TestBlock
 toPoint = castPoint . withOrigin GenesisPoint blockPoint
 
+-- | More efficient implementation of a check used in some of the handlers,
+-- determining whether the first argument is on the chain that ends in the
+-- second argument.
+-- We would usually call @withinFragmentBounds@ for this, but since we're
+-- using 'TestBlock', looking at the hash is cheaper.
+isAncestor ::
+  HasHeader blk1 =>
+  HasHeader blk2 =>
+  HeaderHash blk1 ~ TestHash =>
+  HeaderHash blk2 ~ TestHash =>
+  WithOrigin blk1 ->
+  WithOrigin blk2 ->
+  Bool
+isAncestor (At ancestor) (At descendant) =
+  NonEmpty.isPrefixOf (reverse (NonEmpty.toList hashA)) (NonEmpty.reverse hashD)
```

Comment:
```suggestion
  NonEmpty.isSuffixOf hashA hashD
```

Reply:
Maybe rename `isAncestor` to `isAncestorOf`

Reply:
that function doesn't exist 😅 but we could convert both to `List`, there it's a thing.

Reply:
✅